### PR TITLE
Wasm shrink - "fast exit" serializer strategy

### PR DESCRIPTION
**Pull request by andrei-marinica**
Coded traits now all have "[top|dep]_[encode|decode]_or_exit" variants, which take a context and a reference to an "exit" function.

The exit function acts like a panic, in that it immediately halts execution. This is indicated by the "-> !" return.
It can be specified by the caller, but it will be expanded lower down the call chain. This is not a function pointer since it is not dynamic, it gets expanded statically.
Since encoding never fails in our scenarios, the encoding exit will never appear in smart contracts (unlike the Result<(), EncodeError> which was producing an if-like statement after each return). It is, however, possible to have encoding errors in custom types, so this logic is necessary.
Decoding exit also appears just a few times, because I've pushed it far down the call graph, to the decode input trait implementations, which are called by everyone.
All in all I've managed to chop ~5kB off the basic-features contract (debug info included) with this strategy, compared to the version with closures (v0.8.0).

This serialization pattern seems to be the winner, so it's likely final.

### CODE REVIEW

File: elrond-codec/src/nested_de.rs [hunk 1]
```diff
@@ -71,25 +90,48 @@ impl<T: NestedDecode> NestedDecode for Vec<T> {
...
                 }
                 Ok(result)
 			}
         }
     }
+
+    fn dep_decode_or_exit<I: NestedDecodeInput, ExitCtx: Clone>(input: &mut I, c: ExitCtx, exit: fn(ExitCtx, DecodeError) -> !) -> Self {
+        let size = usize::dep_decode_or_exit(input, c.clone(), exit);
+        match T::TYPE_INFO {
+			TypeInfo::U8 => {
```
Comment (sasurobert) [inline elrond-codec/src/nested_de.rs:103]: indentation

File: elrond-codec/src/nested_de.rs [hunk 2]
```diff
@@ -198,19 +266,31 @@ macro_rules! array_impls {
     ($($n: tt,)+) => {
         $(
             impl<T: NestedDecode> NestedDecode for [T; $n] {
-                fn dep_decode<I: NestedDecodeInput, R, F: FnOnce(Result<Self, DecodeError>) -> R>(input: &mut I, f: F) -> R {
-                    let mut array_vec = ArrayVec::<[T; $n]>::new();
+                fn dep_decode<I: NestedDecodeInput>(input: &mut I) -> Result<Self, DecodeError> {
+					let mut r = ArrayVec::new();
```
Comment (sasurobert) [inline elrond-codec/src/nested_de.rs:270]: indentation

File: elrond-codec/src/nested_de.rs [hunk 3]
```diff
@@ -198,19 +266,31 @@ macro_rules! array_impls {
...
 					for _ in 0..$n {
-                        match T::dep_decode_to(input) {
-                            Ok(elem) => { array_vec.push(elem); }
-                            Err(e) => { return f(Err(e)); }
-                        }
+						r.push(T::dep_decode(input)?);
 					}
-					f(match array_vec.into_inner() {
+					let i = r.into_inner();
```
Comment (sasurobert) [inline elrond-codec/src/nested_de.rs:274]: indentation

File: elrond-codec/src/nested_de.rs [hunk 4]
```diff
@@ -198,19 +266,31 @@ macro_rules! array_impls {
...
 						Ok(a) => Ok(a),
 						Err(_) => Err(DecodeError::ARRAY_DECODE_ERROR),
-					})
-				}
+					}
+                }
+                
+                fn dep_decode_or_exit<I: NestedDecodeInput, ExitCtx: Clone>(input: &mut I, c: ExitCtx, exit: fn(ExitCtx, DecodeError) -> !) -> Self {
+                    let mut r = ArrayVec::new();
+					for _ in 0..$n {
```
Comment (sasurobert) [inline elrond-codec/src/nested_de.rs:284]: indentation

File: elrond-codec/src/nested_de_input.rs [hunk 5]
```diff
@@ -41,6 +57,15 @@ impl<'a> NestedDecodeInput for &'a [u8] {
 		Ok(())
     }
 
+    fn read_into_or_exit<ExitCtx: Clone>(&mut self, into: &mut [u8], c: ExitCtx, exit: fn(ExitCtx, DecodeError) -> !) {
+        if into.len() > self.len() {
+			exit(c, DecodeError::INPUT_TOO_SHORT);
+		}
+		let len = into.len();
```
Comment (sasurobert) [inline elrond-codec/src/nested_de_input.rs:64]: indentation

File: elrond-wasm/src/types/queue.rs [hunk 6]
```diff
@@ -86,37 +86,59 @@ impl<T> Queue<T> {
...
 
 impl<T: NestedEncode> TopEncode for Queue<T> {
+    #[inline]
     fn top_encode<O: TopEncodeOutput>(&self, output: O) -> Result<(), EncodeError> {
         self.as_slice().top_encode(output)
     }
+
+    #[inline]
+    fn top_encode_or_exit<O: TopEncodeOutput, ExitCtx: Clone>(&self, output: O, c: ExitCtx, exit: fn(ExitCtx, EncodeError) -> !) {
+		self.as_slice().top_encode_or_exit(output, c, exit)
```
Comment (sasurobert) [inline elrond-wasm/src/types/queue.rs:106]: indentation

File: examples/lottery/src/status.rs [hunk 7]
```diff
@@ -30,13 +30,23 @@ impl TopEncode for Status {
     fn top_encode<O: TopEncodeOutput>(&self, output: O) -> Result<(), EncodeError> {
         self.to_u8().top_encode(output)
     }
+
+    fn top_encode_or_exit<O: TopEncodeOutput, ExitCtx: Clone>(&self, output: O, c: ExitCtx, exit: fn(ExitCtx, EncodeError) -> !) {
+		self.to_u8().top_encode_or_exit(output, c, exit)
```
Comment (sasurobert) [inline examples/lottery/src/status.rs:35]: indentation

File: test-contracts/basic-features/src/large_boxed_byte_array.rs [hunk 8]
```diff
@@ -7,28 +7,47 @@ pub struct LargeBoxedByteArray(Box<[u8; ARRAY_SIZE]>);
...
     #[inline]
-    fn dep_encode_to<O: NestedEncodeOutput>(&self, dest: &mut O) -> Result<(), EncodeError> {
-        self.0.dep_encode_to(dest)
+    fn dep_encode<O: NestedEncodeOutput>(&self, dest: &mut O) -> Result<(), EncodeError> {
+        self.0.dep_encode(dest)
     }
+
+    #[inline]
+	fn dep_encode_or_exit<O: NestedEncodeOutput, ExitCtx: Clone>(&self, dest: &mut O, c: ExitCtx, exit: fn(ExitCtx, EncodeError) -> !) {
+		self.0.dep_encode_or_exit(dest, c, exit);
```
Comment (sasurobert) [inline test-contracts/basic-features/src/large_boxed_byte_array.rs:16]: indentation

File: test-contracts/basic-features/src/large_boxed_byte_array.rs [hunk 9]
```diff
@@ -7,28 +7,47 @@ pub struct LargeBoxedByteArray(Box<[u8; ARRAY_SIZE]>);
...
 
 impl TopEncode for LargeBoxedByteArray {
     #[inline]
     fn top_encode<O: TopEncodeOutput>(&self, output: O) -> Result<(), EncodeError> {
         self.0.top_encode(output)
     }
+
+    #[inline]
+    fn top_encode_or_exit<O: TopEncodeOutput, ExitCtx: Clone>(&self, output: O, c: ExitCtx, exit: fn(ExitCtx, EncodeError) -> !) {
+		self.0.top_encode_or_exit(output, c, exit);
```
Comment (sasurobert) [inline test-contracts/basic-features/src/large_boxed_byte_array.rs:28]: indentation

File: elrond-codec/src/top_de.rs [hunk 10]
```diff
@@ -26,89 +26,144 @@ pub trait TopDecode: Sized {
...
 }
 
 impl TopDecode for () {
     const TYPE_INFO: TypeInfo = TypeInfo::Unit;
     
-	fn top_decode<I: TopDecodeInput, R, F: FnOnce(Result<Self, DecodeError>) -> R>(_input: I, f: F) -> R {
-		f(Ok(()))
-	}
+	fn top_decode<I: TopDecodeInput>(_: I) -> Result<Self, DecodeError> {
+		Ok(())
```
Comment (sasurobert) [inline elrond-codec/src/top_de.rs:91]: indentation

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: elrond-codec/src/lib.rs [hunk 11]
```diff
@@ -65,34 +65,57 @@ pub mod test_struct {
...
+			self.int.dep_encode(dest)?;
+			self.seq.dep_encode(dest)?;
+            self.another_byte.dep_encode(dest)?;
             Ok(())
-		}
+        }
+        
+        fn dep_encode_or_exit<O: NestedEncodeOutput, ExitCtx: Clone>(&self, dest: &mut O, c: ExitCtx, exit: fn(ExitCtx, EncodeError) -> !) {
+            self.int.dep_encode_or_exit(dest, c.clone(), exit);
+			self.seq.dep_encode_or_exit(dest, c.clone(), exit);
```
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply elrond-codec/src/lib.rs:77]: coming in the next PR

**Review by sasurobert — APPROVED** [review]

**Pull request merged**